src/test/ProposalPayloadE2E.t.sol

```diff
@@ -1,39 +1,238 @@
+    address public constant USDC = 0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48;
+    address public constant USDT = 0xdAC17F958D2ee523a2206206994597C13D831ec7;
+    address public constant WBTC = 0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599;
+    address public constant WETH = 0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2;
+
+    address[5] private ammTokens = [AMMDAI, AMMUSDC, AMMUSDT, AMMWBTC, AMMWETH];
+    address[5] private tokens = [DAI, USDC, USDT, WBTC, WETH];
+
     function setUp() public {
-        // To fork at a specific block: vm.createSelectFork(vm.rpcUrl("mainnet"), BLOCK_NUMBER);
-        vm.createSelectFork(vm.rpcUrl("mainnet"));
+        vm.createSelectFork(vm.rpcUrl("mainnet"), 16183428); // December 14, 2022
+
+        consolidationContract = new AaveV2CollectorContractConsolidation();
+        withdrawContract = new AMMWithdrawer();
+        payload = new ProposalPayload(consolidationContract, withdrawContract);
 
-        // Deploy Payload
-        ProposalPayload proposalPayload = new ProposalPayload();
+        allTokens[0] = consolidationContract.ARAI();
+        allTokens[1] = consolidationContract.AAMPL();
+        allTokens[2] = consolidationContract.AFRAX();
+        allTokens[3] = consolidationContract.FRAX();
+        allTokens[4] = consolidationContract.AENS();
+        allTokens[5] = consolidationContract.SUSD();
+        allTokens[6] = consolidationContract.ASUSD();
+        allTokens[7] = consolidationContract.TUSD();
+        allTokens[8] = consolidationContract.ATUSD();
+        allTokens[9] = consolidationContract.AMANA();
+        allTokens[10] = consolidationContract.ADPI();
+        allTokens[11] = consolidationContract.ABUSD();
+        allTokens[12] = consolidationContract.BUSD();
+        allTokens[13] = consolidationContract.ZRX();
+        allTokens[14] = consolidationContract.AZRX();
+        allTokens[15] = consolidationContract.AUST();
+        allTokens[16] = consolidationContract.MANA();
 
-        // Create Proposal
-        vm.prank(GovHelpers.AAVE_WHALE);
+        vm.startPrank(GovHelpers.AAVE_WHALE);
         proposalId = DeployMainnetProposal._deployMainnetProposal(
-            address(proposalPayload),
-            0x344d3181f08b3186228b93bac0005a3a961238164b8b06cbb5f0428a9180b8a7 // TODO: Replace with actual IPFS Hash
+            address(payload),
+            0x78ce0d63ca0c186ca3f58e712d3f1861ced3dad15ce3ad4f0e005d1663b49caf
         );
+        vm.stopPrank();
     }
 
-    function testExecute() public {
-        // Pre-execution assertations
+    function testWithdrawOfAMMTokens() public {
+        GovHelpers.passVoteAndExecute(vm, proposalId);
+
+        uint256 length = ammTokens.length;
+        for (uint256 i = 0; i < length; i++) {
+            // The withdrawal conversion is not 1 to 1 so might not be able to redeem to 0
+            assertApproxEqAbs(IERC20(ammTokens[i]).balanceOf(AaveV2Ethereum.COLLECTOR), 0, 4 ether);
```

Comment:
Curious as to the 4 ether approximation here. Are there cases where that much is left behind?

Reply:
Can you also add an additional assertion for the underlying asset that's received? And also another assertion to check that no asset is left behind in the AMM Withdrawer either (since you're transferring the aAMMToken into that contract)